src/analyzer/protocol/modbus/modbus-analyzer.pac

```diff
@@ -102,21 +109,21 @@ refine flow ModbusTCP_Flow += {
 		%{
 		// We will assume that if an entire PDU from both sides
 		// is successfully parsed then this is definitely modbus.
-		connection()->SetPDU(${message.is_orig});
+        connection()->SetPDU(${message.is_orig});
 
-		if ( ! connection()->IsConfirmed() )
-			{
-			connection()->SetConfirmed();
-			connection()->zeek_analyzer()->AnalyzerConfirmation();
-			}
+        if ( ! connection()->IsConfirmed() && connection()->IsValidHeader( ${message.header} ) )
+            {
+            connection()->SetConfirmed();
+            connection()->zeek_analyzer()->AnalyzerConfirmation();
+            }
 
 		return true;
 		%}
 
 	# EXCEPTION
 	function deliver_Exception(header: ModbusTCP_TransportHeader, message: Exception): bool
 		%{
-		if ( ::modbus_exception )
+		if ( ::modbus_exception && ( connection()->IsConfirmed() || connection()->IsValidHeader(header) ) )
```

Comment:
I *think* if we use `&enforce` this can go and result in the wanted behavior still.